### ISSUE TITLE
Add support for enumeration (WQL)?

### ISSUE THREAD

**Issue by knalli** [cloudsoft/winrm4j#146]
Hi,

as fas as I understand the used techniques correctly, the service interface [WinRm](https://github.com/cloudsoft/winrm4j/blob/50219d476a874f47d9b7b02f8d92c2a2ddddd09e/service/src/main/java/io/cloudsoft/winrm4j/service/WinRm.java) is the central location for binding the entities respective service endpoints.

I'm unsure how easy it would be adding support for enumeration which is useful for executing WQL queries. Currently, the workaround via a PS command works. However, this requires the startup/instance of a (running) shell which is expensive (both runtime duration and actual cpu performance on the host). Am I too naive and there are some hidden difficulties except for the given five commands?

I've already started experimenting with this but struggling to gather the different api/soap/tokens correctly.

Sidenote: For an actual high-level/end-user solution, there should be a wrapper which merges the result. See also in [WinRb/WinRM](https://github.com/WinRb/WinRM/blob/ebbed119340c665d8220298c0103f0f106b3bc6b/lib/winrm/wsmv/wql_query.rb#L21).

Do you think this a useful addition?

**Comment by AndrewDodd42** [comment]
@knalli I would like to use your work on WQL support and add it into my fork of winrm4j. Would you object to me preparing PRs from your code and submitting them to the origin project?

**Comment by knalli** [comment]
@AndrewDodd42 No, of course not. I'd wished more attention here, actually. :) I have waited for some feedback, but feel free. If you have some questions, feel free to come up. 

We have this running for some time and are quite happy with it. The instrumentation of the underlaying CXF may need some improvements..

**Comment by AndrewDodd42** [comment]
@knalli Great and it's good to know that the WQL code is working well. I'll get started with creating PRs for the smaller fixes that are in your fork and then look at the WQL changes.

I'll probably be back to ask more questions as I'm pretty new to GitHub. What is the best way to collaborate on these changes? Adding comments to this PR?

**Comment by AndrewDodd42** [comment]
@knalli I've created a series of PRs that package up your smaller changes and modify them slightly to get them working against the latest upstream. I've also created a draft PR for the enumeration feature; this appears to work but I want to test it some more tomorrow.

The next stage after that is to attempt to create a PR that depends on both the enumeration and codepage changes and changes buildClient() to work with both of them.

**Comment by knalli** [comment]
Nice to hear. Just ping me.

**Comment by chinaxushi** [comment]
Has it been merged?

**Comment by AndrewDodd42** [comment]
> Has it been merged?

No, I'm not sure the original maintainers are active any longer.

I've put together the WQL changes that knalli made and a few other fixes and improvements in my fork at:- https://github.com/AndrewDodd42/winrm4j

**Comment by knalli** [comment]
Yeah, having also waiting for this. Besides this one, there is also an outstanding review/rework for a compatibility with Jakarta EE (at least since releases of Spring Boot v3 and Apache CXF v4). However, I'm not so in the CXF stuff and I must stopped end of last year. Do you have already progress in this topic?

Maybe we should think about a fork of the additional features?

**Comment by chinaxushi** [comment]
> > Has it been merged?
> 
> No, I'm not sure the original maintainers are active any longer.
> 
> I've put together the WQL changes that knalli made and a few other fixes and improvements in my fork at:- https://github.com/AndrewDodd42/winrm4j

Is it normal when I do the WQL query? The data here seems to return null, but according to the log, the number of returned items seems to be correct

**Comment by bertysentry** [comment]
Hi! 👋

If you're looking for an actively maintained alternative to winrm4j, we maintain **[winrm-java](https://github.com/MetricsHub/winrm-java)**, another WinRM (light) client for Java and its [documentation](https://metricshub.org/winrm-java) with **zero runtime dependencies**, designed to be easy to migrate to from winrm4j:

|  | winrm4j 0.12.x | winrm-java 2.1 |
|---|---|---|
| Dependencies | Apache CXF + HttpClient + JAXB (dozens of jars) | **None** (single ~300 kB jar) |
| Java | 8 (needs manual JAX-WS/JAXB additions on 11+) | 11+ (works out of the box on current JDKs) |
| Commands & PowerShell | `executeCommand` / `executePs` | `command(...)` / `powerShell(...)` — no script size limit, no escaping |
| WQL / WMI queries | — | ✔ (paged, streamable) |
| File transfer | — | ✔ (through the WinRM channel itself, no SMB) |
| Live output, stdin, `Process`-style streaming | partial (Writer tail) | ✔ |
| Maintenance | last commit 2023 | active |

Migration is pretty straightforward:

```java
try (WinRMClient client = WinRMClient.builder("server01").credentials("DOMAIN\\user", password).build()) {
    CommandResult result = client.powerShell("Get-Service | Where-Object Status -eq 'Running'").execute();
    System.out.println(result.stdout());
}
```

There's a [dedicated migration guide](https://metricshub.org/winrm-java/migrating-from-winrm4j.html) mapping every `WinRmTool` / `WinRmClientBuilder` option to its equivalent, including the behavioral differences to watch for

Maven coordinates: `org.metricshub:winrm-java:2.1.00`, and license is Apache-2.0, like winrm4j.
